prow/cmd/build/controller.go

```diff
@@ -17,46 +17,423 @@ limitations under the License.
+}
+
+type reconciler interface {
+	getProwJob(namespace, name string) (*prowjobv1.ProwJob, error)
+	getBuild(namespace, name string) (*buildv1alpha1.Build, error)
+	deleteBuild(namespace, name string) error
+	createBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error)
+	updateBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error)
+	updateProwJob(namespace string, pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error)
+	now() metav1.Time
+}
+
+func (c *controller) getProwJob(namespace, name string) (*prowjobv1.ProwJob, error) {
+	return c.pjc.ProwV1().ProwJobs(namespace).Get(name, metav1.GetOptions{})
+}
+
+func (c *controller) getBuild(namespace, name string) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Get(name, metav1.GetOptions{})
+}
+func (c *controller) deleteBuild(namespace, name string) error {
+	return c.bc.BuildV1alpha1().Builds(namespace).Delete(name, &metav1.DeleteOptions{})
+}
+func (c *controller) createBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Create(b)
+}
+func (c *controller) updateBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Update(b)
+}
+func (c *controller) updateProwJob(namespace string, pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error) {
+	return c.pjc.ProwV1().ProwJobs(namespace).Update(pj)
+}
+
+func (c *controller) now() metav1.Time {
+	return metav1.Now()
+}
+
+var (
+	groupVersionKind = schema.GroupVersionKind{
+		Group:   prowjobv1.SchemeGroupVersion.Group,
+		Version: prowjobv1.SchemeGroupVersion.Version,
+		Kind:    "ProwJob",
+	}
+)
+
+// reconcile ensures that the prowjob and build
+func reconcile(c reconciler, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource %s: %v", key, err))
+	}
+
+	var wantBuild bool
+
+	pj, err := c.getProwJob(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not want build
+	case err != nil:
+		return fmt.Errorf("get prowjob: %v", err)
+	case pj.Spec.Agent != prowjobv1.BuildAgent:
+		// Do not want a build for this job
+	default:
+		wantBuild = true
+	}
+
+	var haveBuild bool
+
+	b, err := c.getBuild(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not have a build
+	case err != nil:
+		return fmt.Errorf("get build: %v", err)
+	default:
+		haveBuild = true
+	}
+
+	// Should we create or delete this build?
+	switch {
+	case !wantBuild:
+		if !haveBuild {
+			logrus.Infof("Deleted %s", key)
+			return nil
+		}
+		switch or := metav1.GetControllerOf(b); {
+		case or == nil, or.APIVersion != groupVersionKind.GroupVersion().String(), or.Kind != groupVersionKind.Kind:
+			return nil // Not controlled by this
+		}
+		logrus.Infof("Delete builds/%s", key)
+		if err = c.deleteBuild(namespace, name); err != nil {
```

Comment:
You probably don't need to delete the build if it's already deleted (b.DeletionTimestamp != nil). In general, since you requeue on deletes, you need to take into account prowjobs and builds in deleted state and handle them appropriately.

Reply:
Done